optuna/storages/_journal/storage.py

```diff
@@ -0,0 +1,608 @@
+            elif op == JournalOperation.SET_STUDY_DIRECTIONS:
+                self._apply_set_study_directions(log)
+            elif op == JournalOperation.CREATE_TRIAL:
+                self._apply_create_trial(log)
+            elif op == JournalOperation.SET_TRIAL_PARAM:
+                self._apply_set_trial_param(log)
+            elif op == JournalOperation.SET_TRIAL_STATE_VALUES:
+                self._apply_set_trial_state_values(log)
+            elif op == JournalOperation.SET_TRIAL_INTERMEDIATE_VALUE:
+                self._apply_set_trial_intermediate_value(log)
+            elif op == JournalOperation.SET_TRIAL_USER_ATTR:
+                self._apply_set_trial_user_attr(log)
+            elif op == JournalOperation.SET_TRIAL_SYSTEM_ATTR:
+                self._apply_set_trial_system_attr(log)
+            else:
+                assert False, "Should not reach."
+
+    def get_study(self, study_id: int) -> FrozenStudy:
+        if study_id not in self._studies:
+            raise KeyError(NOT_FOUND_MSG)
+        return self._studies[study_id]
+
+    def get_all_studies(self) -> List[FrozenStudy]:
+        return list(self._studies.values())
+
+    def get_trial(self, trial_id: int) -> FrozenTrial:
+        if trial_id not in self._trials:
+            raise KeyError(NOT_FOUND_MSG)
+        return self._trials[trial_id]
+
+    def get_all_trials(
+        self, study_id: int, states: Optional[Container[TrialState]]
+    ) -> List[FrozenTrial]:
+        if study_id not in self._studies:
+            raise KeyError(NOT_FOUND_MSG)
+
+        frozen_trials: List[FrozenTrial] = []
+        for trial_id in self._study_id_to_trial_ids[study_id]:
+            trial = self._trials[trial_id]
+            if states is None or trial.state in states:
+                frozen_trials.append(trial)
+        return frozen_trials
+
+    def _raise_if_log_issued_by_pid(self, log: Dict[str, Any], err: Exception) -> None:
+        if log["pid"] == self._pid:
+            raise err
+
+    def _study_exists(self, study_id: int, log: Dict[str, Any]) -> bool:
+        if study_id not in self._studies:
+            self._raise_if_log_issued_by_pid(log, KeyError(NOT_FOUND_MSG))
+            return False
+        return True
+
+    def _apply_create_study(self, log: Dict[str, Any]) -> None:
+        study_name = log["study_name"]
+
+        if study_name in [s.study_name for s in self._studies.values()]:
+            self._raise_if_log_issued_by_pid(
+                log,
+                DuplicatedStudyError(
+                    "Another study with name '{}' already exists. "
+                    "Please specify a different name, or reuse the existing one "
+                    "by setting `load_if_exists` (for Python API) or "
+                    "`--skip-if-exists` flag (for CLI).".format(study_name)
+                ),
+            )
+            return
+
+        study_id = self._next_study_id
+        self._next_study_id += 1
```

Comment:
This actually works. If a worker creates a new study, it will log the operation into the backend. (This is done by the [`_write_log`](https://github.com/wattlebirdaz/optuna/blob/a0cc8dc334a5e9bcea10cc4c6e82ffd4c4131280/optuna/storages/_journal/storage.py#L101) function.) If any worker reads this log from the backend, it will replay it by calling the [`apply_create_study`](https://github.com/wattlebirdaz/optuna/blob/a0cc8dc334a5e9bcea10cc4c6e82ffd4c4131280/optuna/storages/_journal/storage.py#L388) function. (This is done by the [`sync`](https://github.com/wattlebirdaz/optuna/blob/a0cc8dc334a5e9bcea10cc4c6e82ffd4c4131280/optuna/storages/_journal/storage.py#L102) function.) Since the logs are shared by the centralized file, the order of the logs is kept the same, therefore, the all the workers can issue the same study_id for a specific study.

Reply:
Thank you for answering my weird question, I understand. It is in the applying phase.